### PR TITLE
Add unix domain socket transport in netty 4.x via JDK16+

### PR DESCRIPTION
**Pull request by honglh**
### Motivation:

Our use case requires using nio transport for unix-domain-socket (UDS) in netty 4.x. 
Although netty 5.x supports UDS,  it is is not compatible with 4.x. For us and probably many users, having to migrate to 5.x in order to use UDS is not a viable option.

#10991 says it is supported in 5.x and explains the complexity of doing so in 4.x. 
Here, I am proposing a simpler way of adding  to 4.x using JDK16+ NIO transport. 

The simplicity is achieved by allowing `NioServerSocketChannel` to continue to support INET only and use new classes for UDS.

This UDS support in this (4.x) PR is incompatible with 5.x. But since 5.x is incompatible with 4.x anyway, I don't this is is an issue.  It is easy for user to change to use UDS from 5.x when it comes to it.

Thus in summary, this change
1) Simple to add UDS support in 4.x
2) Easy to migrate UDS application to 5.x

### Modification:
Only a few APIs are overloaded to support NIO UDS class.
Most changes are added new files.

### Describe the modifications you've done.
The `NioServerSocketChannel` and `NioSocketChannel` supports only INET channels. In this PR mirroring classes for NIO UDS `NioDomainServerSocketChannel` and `NioDomainSocketChannel` are added.

I modified EchoServer.java to use newly created `NioDomainServerSocketChannel` to show the difference creating a EchoServer with INET socket and with UDS socket.

### Result:
This is a draft PR, so POM changes to use JDK16+ to build netty-transport are not included.
The EchoServer with UDS works as expected and same as it is using INET

Fixes #<GitHub issue number>. 

If there is no issue then describe the changes introduced by this PR.

### CODE REVIEW

**Comment by yawkat** [comment]
@chrisvest @normanmaurer

This patch was created by @honglh with some feedback from me. @honglh works for Oracle, though on a different team than me. They want to get the Micronaut HTTP server (and thus netty) working with domain sockets without epoll.

I will help on cleaning up this PR (commit message, compatibility, javadocs, builds, JDK version checks, CLA…) assuming you agree that the general approach is acceptable in netty 4. While this approach is not compatible with the existing kqueue and epoll domain socket implementations (this patch uses the JDK UnixDomainSocketAddress rather than the netty DomainSocketAddress that kqueue and epoll use), I think this tradeoff is reasonable to keep project layout changes to a minimum. Other than that I have no concerns with this approach.

**Comment by normanmaurer** [comment]
@yawkat let me have a look and come back to you early next week

**Comment by normanmaurer** [comment]
@honglh @yawkat I pushed some changes which I think makes things more correct and reduces some noise. WDYT ?

**Comment by yawkat** [comment]
Thanks @normanmaurer ! Mostly looks fine to me. I've tested it locally and it works, except for the use case we already had a problem with in #13877. When directly constructing a NioDomainServerSocketChannel from a ServerSocketChannel (eg from System.inheritedChannel), and then only registering (not binding) the channel (eg systemd socket activation use case), the channel remains inactive because `bound` is never set to `true`. So we probably again either need a setter or pass in the `bound` value through the constructor.

Another nit: NioDomainServerSocketChannel vs EpollServerDomainSocketChannel, the server and domain are flipped.

**Comment by normanmaurer** [comment]
@yawkat fixed the naming... let me check what we can do about the "bound" stuff here.

**Comment by yawkat** [comment]
@normanmaurer the link at the bottom of https://docs.google.com/forms/d/e/1FAIpQLSfXTK6SnWWFbR50DFhoZq2floCFOUBH3kG8sZP77im5Rknctg/viewform?formkey=dG9wTmhoeGNGd1MtdFdtVXl4TlVSNlE6MQ to change the list of employees is broken

**Comment by normanmaurer** [comment]
@yawkat I just pushed a commit that will allow to correctly detect if a channel was bound before. Also added some tests

**Comment by yawkat** [comment]
You haven't

**Comment by normanmaurer** [comment]
@yawkat 🤦 now I did :D

**Comment by yawkat** [comment]
Thanks! I tried locally, and it works out of the box now. 

Otherwise this PR LGTM, we just need to get the CLA sorted on our end.

**Comment by normanmaurer** [comment]
> Thanks! I tried locally, and it works out of the box now.
> 
> Otherwise this PR LGTM, we just need to get the CLA sorted on our end.

@trustin will fix it.

**Comment by normanmaurer** [comment]
@chrisvest PTAL

**Review by chrisvest — CHANGES_REQUESTED** [review]
Looks good. I had some comments, though.

File: transport/src/main/java/io/netty/channel/socket/nio/NioServerDomainSocketChannel.java [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+    }
+
+    @Override
+    protected void doDisconnect() throws Exception {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    protected int doReadMessages(List<Object> buf) throws Exception {
+        SocketChannel ch = javaChannel().accept();
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply transport/src/main/java/io/netty/channel/socket/nio/NioServerDomainSocketChannel.java:152]: let's just use `SocketUtils`

File: transport/src/main/java/io/netty/channel/socket/nio/NioServerDomainSocketChannel.java [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    }
+
+    @Override
+    protected boolean doWriteMessage(Object msg, ChannelOutboundBuffer in) throws Exception {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    protected void doClose() throws Exception {
+        javaChannel().close();
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply transport/src/main/java/io/netty/channel/socket/nio/NioServerDomainSocketChannel.java:178]: hmm... I would think the jdk implementation would handle all of it.

File: transport/src/main/java/io/netty/channel/socket/nio/NioDomainSocketChannel.java [hunk 3]
```diff
@@ -0,0 +1,651 @@
...
+    }
+
+    @SuppressJava6Requirement(reason = "Usage guarded by java version check")
+    @Override
+    protected SocketAddress localAddress0() {
+        try {
+            return javaChannel().getLocalAddress();
+        } catch (Exception ignore) {
+            // ignore
+        }
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply transport/src/main/java/io/netty/channel/socket/nio/NioDomainSocketChannel.java:290]: I would keep it for now as its more "inline" with what `NioSocketChannel` does

**Comment by trustin** [comment]
Thanks for your patience, @yawkat. Here's the employee list update form: https://docs.google.com/forms/d/e/1FAIpQLScnQObZBItRc5L9EC_DPB3WVdUA7K_uIgC-I7gxqq3G4JGySQ/viewform

**Comment by normanmaurer** [comment]
@yawkat @honglh let me know once you think we can pull this in

File: transport/src/main/java/io/netty/channel/socket/nio/NioServerDomainSocketChannel.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrisvest) [reply transport/src/main/java/io/netty/channel/socket/nio/NioServerDomainSocketChannel.java:179]: I don't think the JDK does anything to the file path after closing a unix domain socket. At least, I can't find anything like that in the JDK source code.

**Comment by yawkat** [comment]
Will do, I've sent the CLA upstairs :)

File: transport/src/main/java/io/netty/channel/socket/nio/NioServerDomainSocketChannel.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (normanmaurer) [reply transport/src/main/java/io/netty/channel/socket/nio/NioServerDomainSocketChannel.java:179]: Interesting... ok let me add it to make it behave the same as the native implementations

[the remark this answers is not in the document]
Comment (normanmaurer) [reply transport/src/main/java/io/netty/channel/socket/nio/NioServerDomainSocketChannel.java:180]: @chrisvest fixed. PTAL again

File: transport/src/main/java/io/netty/channel/socket/nio/NioDomainSocketChannel.java [hunk 4]
```diff
@@ -0,0 +1,651 @@
...
+
+    /**
+     * Create a new instance
+     *
+     * @param parent    the {@link Channel} which created this instance or {@code null} if it was created by the user
+     * @param socket    the {@link SocketChannel} which will be used
+     */
+    public NioDomainSocketChannel(Channel parent, SocketChannel socket) {
+        super(parent, socket);
+        config = new NioDomainSocketChannelConfig(this, socket);
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply transport/src/main/java/io/netty/channel/socket/nio/NioDomainSocketChannel.java:118]: How would you check that ?

**Comment by yawkat** [comment]
CLA should be submitted. Note that the Oracle Point of Contact also changed.

**Review by violetagg — APPROVED** [review]

**Comment by normanmaurer** [comment]
Thanks all !

**Comment by yawkat** [comment]
Thanks!

**Pull request merged**